src/aspire/source/simulation.py

```diff
@@ -468,3 +475,132 @@ def eval_coords(self, mean_vol, eig_vols, coords_est):
+        # Note, relationship with CTF and other sim corruptions still isn't clear to me...
+        if self.noise_adder is None:
+            return np.inf
+
+        # For SNR of Simulations, use the theoretical noise variance
+        # known from the noise_adder instead of deriving from PSD.
+        return super().estimate_snr(noise_power=self.noise_adder.noise_var)
+
+    @classmethod
+    def from_snr(
+        cls,
+        target_snr,
+        *args,
+        sample_n=None,
+        support_radius=None,
+        batch_size=512,
+        signal_power_method="estimate_signal_mean_energy",
+        **kwargs,
+    ):
+        """
+        Generates a Simulation source with a WhiteNoiseAdder
+        configured to produce a target signal to noise ratio.
+
+        :param target_snr: Desired signal to noise ratio of
+            the returned source.
+        :param sample_n: Number of images used for estimate.
+            Defaults to all images in source.
+        :param support_radius: Pixel radius used for masking signal support.
+            Default of None will compute inscribed circle, `self.L // 2`.
+        :param batch_size: Images per batch, defaults 512.
+        :param signal_power_method: Method used for computing signal energy.
+           Defaults to mean via `estimate_signal_mean_energy`.
+           Can use variance method via `estimate_signal_var`.
+        :returns: Simulation source.
+        """
+
+        # Create a Simulation
+        sim = cls(*args, **kwargs)
+
+        # When a user supplies angles and requests `sample_n` subset of images
+        #   randomize the rotations for them to avoid as much bias as possible.
+        # The original rotations stored in `_rots` will be restored later.
+        shuffle = (not sim._uniform_random_angles) and (sample_n is not None)
+        if shuffle:
+            # Store the original rotations.
+            _rots = sim.rotations.copy()
+            # Shuffle the Sim rotations
+            sim.rotations = sim.rotations[np.random.permutation(sim.n)]
+
+        # Assert NoiseAdder has not been provided
+        if "noise_adder" in kwargs or sim.noise_adder is not None:
+            raise RuntimeError(
+                "Cannot provide 'noise_adder' when using {cls.__name__}.from_snr."
+            )
+
+        # Estimate the required noise variance
+        signal_var = sim.estimate_signal_power(
+            sample_n=sample_n,
+            support_radius=support_radius,
+            signal_power_method=signal_power_method,
+        )
+        noise_var = signal_var / target_snr
+
+        # Assign the noise_adder
+        sim.noise_adder = WhiteNoiseAdder(var=noise_var)
+        logger.info(f"Appended {sim.noise_adder} to generation pipeline")
+
+        # Restore the original rotations when previously shuffled.
+        if shuffle:
+            sim.rotations = _rots
+
+        return sim
+
+    def estimate_psnr(self, sample_n=None, batch_size=512, units=None):
```

Comment:
Do we really need this? I haven't seen any PSNR measures for cryo-EM methods. Seems more common in the wider image/signal processing literature.

Reply:
I will remove.  I think it might come in handy for other situations later in our pipeline, but I'll just keep it in my back pocket until then...